### ISSUE TITLE
leafref type is not working inside union type

### ISSUE THREAD

**Issue by lioyerar** [CESNET/libyang#1927]
Hello, 
I encountered an issue when defining a leaf whose type is a union that contains only leafref types, then I only added a value for this leaf, the lyd_parse_mem return SUCCESS.
I have just compiled the newest libyang and tried the following YANG schema in the file vrouter-snmp.yang:

	module vrouter-snmp {
	  yang-version 1.1;
	  namespace "urn:test:vrouter/snmp";
	  prefix vrouter-snmp;
	  organization
	    "test";
	  contact
	    "test support - <support@test.com>";
	  description
	    "test.";

	grouping snmp {
	    description
	      "SNMP common options.";

	list view {
      key "name";
      description
        "A named 'view'.";

      leaf name {
        type int64;
        description
          "The name of the view.";
      }
    }

    list view2 {
      key "name";
      description
        "A named 'view'.";

      leaf name {
        type int64;
        description
          "The name of the view.";
	   }
	}

    // list view
    list community {
      key "name";
      description
        "An SNMPv1 or SNMPv2c community.";

      leaf name {
        type string;
        description
          "The name of the community.";
      }

      leaf view {
        type union {
          type leafref {
            path
              "../../view/name";
            require-instance true;
          }
          type leafref {
            path
              "../../view2/name";
            require-instance true;
          }
        }
        description
          "test.";
      }
      leaf view2 {
        type leafref {
          path
            "../../view/name";
	        }
	      }
	    }
	  }
	  container test {
		  description "test";
	  uses snmp;
	  }
	}

the test program:

	#include <stdio.h>
	#include <libyang/libyang.h>
	
	int main (void){
        struct ly_ctx *ctx = NULL;
				struct lyd_node *root = NULL;
        char *xml =
        "<test xmlns=\"urn:test:vrouter/snmp\">"
                "<community>"
                        "<name>test</name>"
                        "<view>111111</view>"
                "</community>"
        "</test>";

        char *xml2 =
        "<test xmlns=\"urn:test:vrouter/snmp\">"
                "<community>"
                        "<name>test</name>"
                        "<view>111111</view>"
                        "<view2>2222</view2>"
                "</community>"
        "</test>";

        ly_ctx_new(NULL, 0, &ctx);
        lys_parse_path(ctx, "vrouter-snmp.yang", LYS_IN_YANG, NULL);
        
        /* ok */
        if(lyd_parse_data_mem(ctx, xml, LYD_XML, LYD_PARSE_STRICT, LYD_VALIDATE_PRESENT, &root) == LY_SUCCESS)
                lyd_print_file(stdout, root, LYD_XML, 0);

        /* fail */
        if(lyd_parse_data_mem(ctx, xml2, LYD_XML, LYD_PARSE_STRICT, LYD_VALIDATE_PRESENT, &root) == LY_SUCCESS)
                lyd_print_file(stdout, root, LYD_XML, 0);

        return 0;
      }

The output is:

	lsang@disturbed:~$gcc union_test.c -lyang  -Wall -Werror -Wextra && ./a.out 
	<test xmlns="urn:test:vrouter/snmp">
	  <community>
    <name>test</name>
    <view>111111</view>
	  </community>
	</test>
	libyang[0]: Invalid leafref value "2222" - no target instance "../../view/name" with the same value. (path: Schema location "/vrouter-snmp:test/community/view2", data location "/vrouter-snmp:test/community[name='test']/view2".)

The first call of lyd_parse_data_mem returned LY_SUCCESS whereas neither list view nor list pouet is defined.
The second call of lyd_parse_data_mem worked as expected.

I do not have reproduced this with libyang version 1.

**Comment by michalvasko** [comment]
Right, leafref targets in unions were not being searched for, thanks for reporting this. Fixed, test added.

**Comment by lioyerar** [comment]
Thanks for the fix, I will test on my side.

**Comment by lioyerar** [comment]
Hello,   

I have taken your patch and tested it on my side. The simple test program above is now working as expected.
However, I encountered another problem that I am not sure is related to libyang, Netopeer or sysrepo.

With the almost identical schema vrouter-snmp.yang,  instead of defining a list element in view, I defined a list element in view2.
The test program is working as expected, but if I send a rpc with netopper-cli, it returned an error with the message "User callback failed.

Example:
with netopeer2-cli:
 > get-schema --model vrouter-snmp


	module vrouter-snmp {

	yang-version  1.1;

	namespace  "urn:6wind:vrouter/snmp";

	prefix vrouter-snmp;
	
	...	
		list view {
	      key "name";
	      description
	        "A named 'view' - a subset of the overall OID tree.";
	      leaf name {
	        type string;
	        description
	          "The name of the view.";
	      }
	      list subtree {
	        ext:nc-cli-one-liner;
	        key "oid";
	        min-elements 1;
	        description
	          "A part of the OID tree to include or exclude from the view.";  
	        leaf oid {
	          type oid;
	          description
	            "The OID root to include or exclude from the view.";          
	        }
	        leaf included {
	          type boolean;
	          default "true";
	          description
	            "Set to false to exclude this OID from the view.";            
	        }
	      }
	    }
    
	    list view2 {
	      key "name";
	      description
	        "A named 'view' - a subset of the overall OID tree.";
	      leaf name {
	        type string;
	        description
	        "The name of the view.";
	      }
	      list subtree {
	        ext:nc-cli-one-liner;
	        key "oid";
	        min-elements 1;
	        description
	          "A part of the OID tree to include or exclude from the view.";  
	        leaf oid {
	          type oid;
	          description
            "	The OID root to include or exclude from the view.";          
	        }
	        leaf included {
	          type boolean;
	          default "true";
	          description
	            "Set to false to exclude this OID from the view.";            
	        }
	      }
	    }
    
	    list community {
	      key "name";
	      description
	        "An SNMPv1 or SNMPv2c community.";
	      leaf name {
	        type string;
	        description
	          "The name of the community.";
	      }
	      leaf authorization {
	        type auth-level;
	        mandatory true;
	        description
	          "The authorization level of the community.";
	      }
	      leaf-list source {
	        type union {
	          type vr-inet:host;
	          type vr-inet:ip-prefix;
	        }
	        description
	          "Restrict access to requests from the specified address or prefix list."; 
	      }
	      leaf view {
	        type union {
	          type leafref {
	            path "../../view/name";
	          }
	          type leafref {
	            path "../../view2/name";
	          }
	        }
	        description
	          "Restricts access for that community to the subtree rooted at the         
	           given view name. If not specified, the community has access to the       
	           whole OID tree.";
	      }
	  ...
	}

No data tree yet for vrouter-snmp:
> get-data  --datastore running  --filter-xpath "/vrouter:config/vrf[name='main']/vrouter-snmp:snmp"

    DATA
    \<data xmlns="urn:ietf:params:xml:ns:yang:ietf-netconf-nmda"/>
    

I try to edit the config with this XML file, rpc.xml:

	   
	<config xmlns="urn:6wind:router">
	     <vrf>
	      <name>main</name>
	    <snmp xmlns="urn:6wind:vrouter/snmp">
	      <enabled>true</enabled>
	      <view2>
	        <name>test</name>
	        <subtree>
	          <oid>1.1.1</oid>
	          <included>true</included>
	        </subtree>
	      </view2>
	      <community>
	        <name>test</name>
	        <authorization>read-only</authorization>
	        <view>test</view>
	      </community>
	    </snmp>
	  </vrf>
	</config>

> edit-config  --target running --config=rpc.xml --defop replace



	ERROR
        type:     application
        tag:      operation-failed
        severity: error
        message:  Invalid leafref value "test" - no target instance "../../view/name" with the same value, target path is ../../view[name='test']/name. (Schema location "/vrouter:config/vrf/vrouter-snmp:snmp/community/view", data location "/vrouter
	:config/vrf[name='main']/vrouter-snmp:snmp/community[name='test']/view".)                                                  

        type:     application
        tag:      operation-failed
        severity: error
        message:  User callback failed.

===> can not apply.

Whereas, if I try with sysrepocfg:

>root@dut-vm:~# sysrepocfg --edit=rpc.xml
 
And verify in netopeer2-cli:


> get-data  --datastore running  --filter-xpath "/vrouter:config/vrf[name='main']/vrouter-snmp:snmp" 

	DATA
	<data xmlns="urn:ietf:params:xml:ns:yang:ietf-netconf-nmda">
	  <config xmlns="urn:6wind:vrouter">
	    <vrf>
	      <name>main</name>
	      <snmp xmlns="urn:6wind:vrouter/snmp">
	        <enabled>true</enabled>
	        <view2>
	          <name>test</name>
	          <subtree>
	            <oid>1.1.1</oid>
	            <included>true</included>
	          </subtree>
	        </view2>
	        <community>
	          <name>test</name>
	          <authorization>read-only</authorization>
	          <view>test</view>
	        </community>
	      </snmp>
	    </vrf>
	  </config>
	</data>

And validate the running datasource:
> validate  --source running

    OK


The first remark comes to me, it may be in the union plugin, but there are still some errors logs that are not cleaned, I did some debugging with ly_errmsg, and it returned NULL.

Do you have an idea?
Thanks

**Comment by michalvasko** [comment]
Should not be difficult to understand and fix but please provide some usable YANG modules corresponding to the data, the single snippet is not enough, it seems it is just an augment of the base vrouter YANG module.

**Comment by lioyerar** [comment]
Hello, 

I have updated my previous comment with a test YANG module.

Thanks.

**Comment by michalvasko** [comment]
Sorry, I have not tested it with LYB values, should now work for them as well.
